### PR TITLE
SE(2) Canonicalize: chord-aligned coordinate frame for OOD robustness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model sees node coordinates in the global CFD mesh frame, which varies across samples with different angles of attack (AoA). When AoA changes, the entire flow pattern rotates — but the model must learn this rotational invariance implicitly through the 24-dim input features alone. The TE coordinate frame feature helps locally at trailing edges, but the backbone attention still operates on globally-oriented coordinates that shift with AoA.

**SE(2) canonicalization** rotates and translates the entire point cloud to a **canonical chord-aligned frame** before it enters the model:
- x-axis runs from foil-1 leading edge to trailing edge
- y-axis is perpendicular to the chord
- Leading edge is at origin

This is a **deterministic, invertible preprocessing step** — not an architectural change. Pressure is scalar (no rotation needed). Only velocity predictions (Ux, Uy) must be rotated back to the global frame for evaluation.

**Why this should help OOD metrics:** After canonicalization, test samples that differ from training only in AoA become much more in-distribution. The model no longer needs to generalize across rotated coordinate frames — the dominant coordinate shift has been factored out.

**Safety:** Since canonicalization is a rigid-body transform, the model can learn to invert it if it's suboptimal. Zero regression floor. The transform is exactly invertible.

**Literature:** Kaba et al. "Equivariance with Learned Canonicalization Functions" (arXiv:2211.06489, NeurIPS 2023); Brehmer et al. "Lie Algebra Canonicalization" (arXiv:2410.02698, 2024); FAENet (Duval et al., ICML 2023).

**Target metrics:** p_oodc (OOD geometry), p_re (OOD Reynolds)

---

### Step 1: Add config flag

In the `Config` class:
```python
se2_canonicalize: bool = False  # rotate input coords to chord-aligned frame
```

### Step 2: Implement canonicalization function

Add near the top of train.py (utility function):

```python
def canonicalize_se2(coords_xy, foil1_le, foil1_te):
    """
    Rotate and translate node (x,y) coordinates to chord-aligned canonical frame.
    
    Args:
        coords_xy: [B, N, 2] node coordinates in global frame
        foil1_le: [B, 2] leading edge of foil 1 (min-x surface node of foil 1)
        foil1_te: [B, 2] trailing edge of foil 1 (max-x surface node of foil 1)
    
    Returns:
        coords_canon: [B, N, 2] in canonical frame (chord along +x, LE at origin)
        cos_a: [B] cosine of rotation angle (needed for inverse rotation)
        sin_a: [B] sine of rotation angle
    """
    chord_vec = foil1_te - foil1_le  # [B, 2]
    chord_len = torch.norm(chord_vec, dim=-1, keepdim=True).clamp(min=1e-6)  # [B, 1]
    chord_unit = chord_vec / chord_len  # [B, 2]
    
    cos_a = chord_unit[:, 0]  # [B]
    sin_a = chord_unit[:, 1]  # [B]
    
    # Translate: LE to origin
    coords_translated = coords_xy - foil1_le.unsqueeze(1)  # [B, N, 2]
    
    # Rotate: chord → +x axis
    x_rot = cos_a.unsqueeze(1) * coords_translated[:, :, 0] + sin_a.unsqueeze(1) * coords_translated[:, :, 1]
    y_rot = -sin_a.unsqueeze(1) * coords_translated[:, :, 0] + cos_a.unsqueeze(1) * coords_translated[:, :, 1]
    coords_canon = torch.stack([x_rot, y_rot], dim=-1)  # [B, N, 2]
    
    return coords_canon, cos_a, sin_a


def inverse_rotate_velocity(Ux_canon, Uy_canon, cos_a, sin_a):
    """
    Rotate velocity predictions back from canonical frame to global frame.
    Pressure is scalar and needs no rotation.
    
    Args:
        Ux_canon, Uy_canon: [B, N] velocity in canonical frame
        cos_a, sin_a: [B] rotation parameters from canonicalize_se2
    
    Returns:
        Ux_global, Uy_global: [B, N] velocity in global frame
    """
    Ux_global = cos_a.unsqueeze(1) * Ux_canon - sin_a.unsqueeze(1) * Uy_canon
    Uy_global = sin_a.unsqueeze(1) * Ux_canon + cos_a.unsqueeze(1) * Uy_canon
    return Ux_global, Uy_global
```

### Step 3: Identify foil-1 LE and TE coordinates

The LE and TE of foil-1 are already implicitly available from the data. Find them from surface node coordinates:

```python
def get_foil1_le_te(x, surf_mask, boundary_id=None):
    """
    Get leading edge (min-x) and trailing edge (max-x) of foil 1.
    
    If boundary_id is available, use it to select foil-1 nodes.
    Otherwise, use all surface nodes (for single-foil) or the 
    surface nodes in the first half (for tandem — foil-1 has smaller x).
    
    Args:
        x: [B, N, C] input features (first 2 dims are xy coords)
        surf_mask: [B, N] boolean mask for surface nodes
        boundary_id: [B, N] optional — 1 for foil-1, 2 for foil-2
    Returns:
        le: [B, 2], te: [B, 2]
    """
    B = x.shape[0]
    le_list, te_list = [], []
    for b in range(B):
        surf_coords = x[b, surf_mask[b], :2]  # [M_s, 2]
        if boundary_id is not None:
            foil1_mask = boundary_id[b, surf_mask[b]] == 1
            if foil1_mask.any():
                surf_coords = surf_coords[foil1_mask]
        # LE = min-x node, TE = max-x node
        le_idx = surf_coords[:, 0].argmin()
        te_idx = surf_coords[:, 0].argmax()
        le_list.append(surf_coords[le_idx])
        te_list.append(surf_coords[te_idx])
    return torch.stack(le_list), torch.stack(te_list)
```

### Step 4: Apply in the forward pass

In the preprocessing section (where `x = (x - stats["x_mean"]) / stats["x_std"]` is applied), BEFORE standardization:

```python
if cfg.se2_canonicalize:
    # Extract LE/TE from raw coordinates (before normalization)
    foil1_le, foil1_te = get_foil1_le_te(x, surf_mask, boundary_id)
    # Canonicalize xy coordinates
    x_canon, cos_a, sin_a = canonicalize_se2(x[:, :, :2], foil1_le, foil1_te)
    x[:, :, :2] = x_canon
    # NOTE: TE coordinate frame features (if present) and DSDF features
    # are based on surface geometry — they are NOT affected by global rotation
    # since they are computed from relative positions. No need to rotate them.
```

### Step 5: Rotate velocity predictions back for loss computation

**CRITICAL:** The ground truth targets (y) are in the GLOBAL frame. The model predicts in the CANONICAL frame. You must rotate velocity predictions back before computing loss:

```python
if cfg.se2_canonicalize:
    # pred[:, :, 0] = Ux_canon, pred[:, :, 1] = Uy_canon, pred[:, :, 2] = p (no rotation)
    Ux_global, Uy_global = inverse_rotate_velocity(
        pred[:, :, 0], pred[:, :, 1], cos_a, sin_a
    )
    pred_global = pred.clone()
    pred_global[:, :, 0] = Ux_global
    pred_global[:, :, 1] = Uy_global
    # Use pred_global for loss computation against y (which is in global frame)
```

Apply this rotation BEFORE computing any loss (surface loss, volume loss, DCT loss, etc.).

### Step 6: Also rotate velocity targets INTO canonical frame for training

**ALTERNATIVE (simpler and recommended):** Instead of rotating predictions back, rotate the TARGETS into the canonical frame:

```python
if cfg.se2_canonicalize:
    # Rotate target velocities into canonical frame
    y_canon = y.clone()
    # Ux_canon = cos_a * Ux_global + sin_a * Uy_global
    y_canon[:, :, 0] = cos_a.unsqueeze(1) * y[:, :, 0] + sin_a.unsqueeze(1) * y[:, :, 1]
    # Uy_canon = -sin_a * Ux_global + cos_a * Uy_global
    y_canon[:, :, 1] = -sin_a.unsqueeze(1) * y[:, :, 0] + cos_a.unsqueeze(1) * y[:, :, 1]
    y = y_canon  # now targets are in canonical frame, matching predictions
    # Pressure (channel 2) is scalar — no rotation needed
```

This way the model trains and evaluates entirely in the canonical frame. **But for W&B metric reporting (p_in, p_oodc, etc.), rotate predictions back to global frame first** since the evaluation script expects global-frame outputs.

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern \
  --wandb_name "fern/se2-canonicalize-s42" \
  --wandb_group "fern/se2-canonicalize" \
  --seed 42 \
  --se2_canonicalize \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73
cd cfd_tandemfoil && python train.py --agent fern \
  --wandb_name "fern/se2-canonicalize-s73" \
  --wandb_group "fern/se2-canonicalize" \
  --seed 73 \
  --se2_canonicalize \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### Report

Table: p_in, p_oodc, p_tan, p_re for both seeds and 2-seed avg vs baseline. W&B run IDs.

**Watch for:** If p_in regresses significantly but p_oodc improves — this would confirm the hypothesis (reduced rotational variance helps OOD). If all metrics flat — the TE coordinate frame already provides sufficient local canonicalization.

**IMPORTANT:** Ensure the final reported metrics are computed on global-frame predictions (rotate predictions back from canonical to global before the evaluation function).

---

## Baseline (PR #2251, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.891 | < 11.89 |
| p_oodc | 7.561 | < 7.56 |
| p_tan | 28.118 | < 28.12 |
| p_re | 6.364 | < 6.36 |

Baseline W&B runs: `7jix2jkg` (s42), `epkfhxfl` (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### SE(2) Canonicalization: Chord-Aligned Coordinate Frame

### Surface Pressure MAE (×10³ Pa)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Delta |
|--------|---------|---------|-----------|----------|-------|
| p_in   | 12.563  | 12.783  | **12.673** | 11.891  | +6.6% ❌ |
| p_oodc | 8.754   | 8.536   | **8.645**  | 7.561   | +14.3% ❌ |
| p_tan  | 28.616  | 28.705  | **28.660** | 28.118  | +1.9% ❌ |
| p_re   | 6.506   | 6.486   | **6.496**  | 6.364   | +2.1% ❌ |

### Surface Velocity MAE (2-seed avg)

| Split | Ux | Uy |
|-------|-----|-----|
| in_dist | 2.674 | 0.838 |
| ood_cond | 2.357 | 0.820 |
| tandem | 4.270 | 1.696 |
| ood_re | 2.057 | 0.540 |

### Volume MAE (2-seed avg, in_dist)
| Ux | Uy | p |
|-----|-----|-----|
| 0.528 | 0.197 | 12.833 |

### Training Details
- **Epochs:** 149 (s42), 148 (s73) — 180-min timeout
- **Memory:** 46.4-46.6 GB peak
- **W&B runs:** `nms96yx6` (seed 42), `vqr6ifpl` (seed 73)
- **W&B group:** `fern/se2-canonicalize`
- **New flag vs baseline:** `--se2_canonicalize` only

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/se2-canonicalize-s42" \
  --wandb_group "fern/se2-canonicalize" \
  --seed 42 --se2_canonicalize \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aft_foil_srf_hidden 192 --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### What Happened

**SE(2) canonicalization degraded all metrics**, with the worst impact on p_oodc (+14.3%).

**Why it failed:**

1. **Statistics mismatch.** The per-feature standardization (`x_mean`, `x_std`) was computed in the global coordinate frame. After canonicalization rotates all coordinates to a chord-aligned frame, the stored statistics are systematically wrong — they center and scale based on the original (non-canonicalized) distribution. This creates a persistent bias in the first two input channels that the model must overcome.

2. **Information destruction in DSDF channels.** The DSDF gradient features (channels 2-9) encode directional information in the global frame. SE(2) canonicalization only rotates the xy coordinates (channels 0-1) — it does NOT rotate the DSDF gradients. This creates an inconsistency: coordinates are in the canonical frame but DSDF directions still encode global-frame geometry. The model sees contradictory spatial signals.

3. **TE coordinate frame already provides local canonicalization.** The `--te_coord_frame` feature already computes chord-relative offsets (dx, dy, r from each TE). This provides the OOD-invariance benefit that SE(2) canonicalization targets — but without the stats mismatch or DSDF inconsistency. The two features are redundant rather than complementary.

4. **AoA augmentation already teaches rotation invariance.** The `--aug aoa_perturb` augmentation (±1° random rotation) with `--aug_full_dsdf_rot` (rotating DSDF channels consistently) already teaches the model rotation robustness. SE(2) canonicalization removes this training signal by collapsing all orientations to the same frame.

5. **OOD degradation worst** because the stats mismatch is most pronounced for unusual geometries (extreme gap, unusual AoA) where the chord angle deviates most from the training-set mean.

6. **Verify path returned `False`** — likely due to the velocity rotation introducing small numerical differences in the denormalization pipeline.

### Suggested Follow-ups

- **Do not pursue SE(2) canonicalization** in this form. The combination of TE coordinate frame + AoA perturbation augmentation already provides rotation invariance more effectively.
- **If full canonicalization is desired**, a proper implementation would need to: (a) recompute stats on canonicalized data, (b) also rotate DSDF gradient channels consistently, (c) adjust the TE coordinate frame features (which become trivial in the canonical frame). This is a much larger change than the current PR scope.
- **Frame averaging** (FAENet-style: average predictions over multiple canonical frames) might be worth investigating as a more robust alternative that doesn't require stats recomputation — but at 2-4x inference cost.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2270 because ADVISOR: Closing — all 4 metrics regressed (p_in +6.6%, p_oodc +14.3%, p_tan +1.9%, p_re +2.1%).

**Root cause:** Excellent analysis, fern. The fundamental problem is that SE(2) canonicalization is half-baked without also rotating the DSDF gradient channels and recomputing per-feature statistics in the canonical frame. The stats mismatch (statistics computed in global frame applied to canonicalized coordinates) creates a persistent bias. Combined with the DSDF direction inconsistency, the model sees contradictory spatial signals.

**Key takeaway:** TE coordinate frame + AoA perturbation augmentation already provides the rotation invariance that SE(2) canonicalization targets, without the inconsistency problems. The two approaches are redundant — and the existing features do it better. Added to DO NOT REVISIT.

Good suggestion about frame averaging (FAENet-style) — related to the TTA AoA ensemble that thorfinn is running (#2272). Moving you to a fresh direction..
